_posts/2020-11-12-deconvolution.md

```diff
@@ -0,0 +1,436 @@
+  <line x1="36" y1="26" x2="156" y2="26" />
+  <line x1="42" y1="32" x2="162" y2="32" style="stroke-width:2" />
+
+  <!-- Vertical lines -->
+  <line x1="10" y1="0" x2="42" y2="32" style="stroke-width:2" />
+  <line x1="40" y1="0" x2="72" y2="32" />
+  <line x1="70" y1="0" x2="102" y2="32" />
+  <line x1="100" y1="0" x2="132" y2="32" />
+  <line x1="130" y1="0" x2="162" y2="32" style="stroke-width:2" />
+
+  <!-- Colored Rectangle -->
+  <polygon points="10.0,0.0 130.0,0.0 162.74356617647058,32.74356617647059 42.74356617647059,32.74356617647059" style="fill:#ECB172A0;stroke-width:0"/>
+
+  <!-- Horizontal lines -->
+  <line x1="42" y1="32" x2="162" y2="32" style="stroke-width:2" />
+  <line x1="42" y1="62" x2="162" y2="62" />
+  <line x1="42" y1="92" x2="162" y2="92" />
+  <line x1="42" y1="122" x2="162" y2="122" />
+  <line x1="42" y1="152" x2="162" y2="152" style="stroke-width:2" />
+
+  <!-- Vertical lines -->
+  <line x1="42" y1="32" x2="42" y2="152" style="stroke-width:2" />
+  <line x1="72" y1="32" x2="72" y2="152" />
+  <line x1="102" y1="32" x2="102" y2="152" />
+  <line x1="132" y1="32" x2="132" y2="152" />
+  <line x1="162" y1="32" x2="162" y2="152" style="stroke-width:2" />
+
+  <!-- Colored Rectangle -->
+  <polygon points="42.74356617647059,32.74356617647059 162.74356617647058,32.74356617647059 162.74356617647058,152.74356617647058 42.74356617647059,152.74356617647058" style="fill:#ECB172A0;stroke-width:0"/>
+
+  <!-- Text -->
+  <text x="102.743566" y="172.743566" font-size="1.0rem" font-weight="100" text-anchor="middle" >2048</text>
+  <text x="182.743566" y="92.743566" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(-90,182.743566,92.743566)">2048</text>
+  <text x="16.371783" y="156.371783" font-size="1.0rem" font-weight="100" text-anchor="middle" transform="rotate(45,16.371783,156.371783)">950</text>
+</svg>
+</td>
+</tr>
+</table>
+
+
+What we now have is a Dask array composed of 16 CuPy blocks of data.  Notice
+how Dask provides nice typing information in the SVG output.  When we moved
+from NumPy to CuPy, the block diagram above displays `Type: cupy.ndarray` --
+this is a nice sanity check.
+
+The last piece we need before running the deconvolution is the PSF which should
+also be loaded onto the GPU:
+
+```python
+psf = skimage.io.imread("/public/NVMICROSCOPY/PSF.tif")
+c_psf = cp.asarray(psf)
+```
+
+Lastly, we call `map_overlap` with the `deconvolve` function across the Dask
+array:
+
+```python
+out = da.map_overlap(
+    deconvolve,
+    c_imgs,
+    psf=c_psf,
+    iterations=20,
+    meta=c_imgs._meta,
+    depth=tuple(np.array(c_psf.shape) // 2),
+    boundary="periodic"
+)
+out
+```
+
+<a href="/images/deconvolve.png">
+    <img src="/images/deconvolve.png" width="100%"></a>
+
+The image above is taken from a mouse intestine.
+
+With Dask and multiple GPUs, we measured deconvolution of an 8GB image in ~10
+seconds! But this is just the first step towards accelerated image science.
```

Comment:
It would be nice to have a reference for CPU performance to compare, even if it is just for 1 or 2 iterations. Not sure what the CPU configuration is in the DGX2?

Reply:
Here's the [specs for a DGX-2]( https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/dgx-2/dgx-2-print-datasheet-738070-nvidia-a4-web-uk.pdf ).

Basically we have 16 Tesla V100s and 24 Dual Intel Xeon Platinum 8168. I guess this raises a question of how we compare these. Do we just use all GPUs and compare to all CPUs? Should we use an equal number of both? Maybe we should do something different?

cc @quasiben (in case you have thoughts here as well)